MIR: opt-in normalization of `BasicBlock` and `Local` numbering

scottmcm opened the pull request:
This doesn't matter at all for actual codegen, but after spending some time reading pre-codegen MIR, I was wishing I didn't have to jump around so much in reading post-inlining code.

So this add two passes that are off by default for every mir level, but can be enabled (`-Zmir-enable-passes=+ReorderBasicBlocks,+ReorderLocals`) for humans.

scottmcm committed:
Tweak the post-order for multi-successor blocks

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5036743874/jobs/9033116545) [(plain)](https://github.com/rust-lang/rust/commit/4533b0a1c3f3f712ef76146b36d1db9912bca421/checks/13637471655/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:8e5e7e5ab8b370d6c329ec480221332ada57f0ab)
Download action repository 'rust-lang/simpleinfra@master' (SHA:af9a3f1b7e12a54c737d8aa371acc8d05cb83a8f)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180115 sha256=c3e1cf75e5aea7348ad201d16633573bc9f3aa58c8e29a6ca459f306300d6c9d
  Stored in directory: /tmp/pip-ephem-wheel-cache-crotg9do/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
Successfully tagged rust-ci:latest
Built container sha256:95c567d2b3a0a9798f2c6b2b40760075136d4d025da5b998a7ad9bf87adc6302
Uploading finished image to https://ci-caches.rust-lang.org/docker/e51156f19850ce886cec818e46dc2a021e0aa7c270d15673e8fe74cd8522fc8ac3995109aebb688ee49ed586735ed4cf5f8c06d44c48298fa09c35ae4b082281

<botocore.awsrequest.AWSRequest object at 0x7fae8e8dbe50>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
##[endgroup]
fmt check
Diff in /checkout/compiler/rustc_mir_transform/src/prettify.rs at line 6:
 
 use crate::MirPass;
 use rustc_index::{bit_set::BitSet, IndexSlice, IndexVec};
+use rustc_middle::mir::visit::{MutVisitor, PlaceContext, Visitor};
 use rustc_middle::mir::*;
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_mir_transform/src/dest_prop.rs" "/checkout/compiler/rustc_mir_transform/src/large_enums.rs" "/checkout/compiler/rustc_mir_transform/src/early_otherwise_branch.rs" "/checkout/compiler/rustc_mir_transform/src/elaborate_box_derefs.rs" "/checkout/compiler/rustc_mir_transform/src/remove_storage_markers.rs" "/checkout/compiler/rustc_mir_transform/src/prettify.rs" "/checkout/compiler/rustc_mir_transform/src/remove_noop_landing_pads.rs" "/checkout/compiler/rustc_mir_transform/src/errors.rs"` failed.
 use rustc_middle::ty::TyCtxt;
-use rustc_middle::mir::visit::{MutVisitor, PlaceContext, Visitor};
 
 pub struct ReorderBasicBlocks;
Diff in /checkout/compiler/rustc_mir_transform/src/prettify.rs at line 25:
             return;
             return;
         }
 
-        let mut updater = BasicBlockUpdater {
-            map: rpo.invert_bijective_mapping(),
-        };
-        };
+        let mut updater = BasicBlockUpdater { map: rpo.invert_bijective_mapping(), tcx };
         debug_assert_eq!(updater.map[START_BLOCK], START_BLOCK);
         updater.visit_body(body);
Diff in /checkout/compiler/rustc_mir_transform/src/prettify.rs at line 44:
     }
 
 
     fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
-        let mut finder = LocalFinder {
-            map: IndexVec::new(),
-            seen: BitSet::new_empty(body.local_decls.len()),
+        let mut finder =
+        let mut finder =
+            LocalFinder { map: IndexVec::new(), seen: BitSet::new_empty(body.local_decls.len()) };
         // We can't reorder the return place or the arguments
         // We can't reorder the return place or the arguments
         for i in 0..=body.arg_count {
Diff in /checkout/compiler/rustc_mir_transform/src/prettify.rs at line 68:
         }
 
 
-        let mut updater = LocalUpdater {
-            map: finder.map.invert_bijective_mapping(),
-        };
-        };
+        let mut updater = LocalUpdater { map: finder.map.invert_bijective_mapping(), tcx };
         debug_assert_eq!(updater.map[RETURN_PLACE], RETURN_PLACE);
         updater.visit_body_preserves_cfg(body);
Diff in /checkout/compiler/rustc_mir_transform/src/prettify.rs at line 98:
         self.tcx
     }
 
 
-    fn visit_terminator(
-        &mut self,
-        terminator: &mut Terminator<'tcx>,
-        _location: Location
-    ) {
+    fn visit_terminator(&mut self, terminator: &mut Terminator<'tcx>, _location: Location) {
         for succ in terminator.successors_mut() {
             *succ = self.map[*succ];
Diff in /checkout/compiler/rustc_mir_transform/src/prettify.rs at line 125:
Diff in /checkout/compiler/rustc_mir_transform/src/prettify.rs at line 125:
 impl<'tcx> Visitor<'tcx> for LocalFinder {
     fn visit_local(&mut self, l: Local, context: PlaceContext, _location: Location) {
         if context.is_use() {
-           self.track(l);
+            self.track(l);
     }
 }
```

</details>

cjgillot commented on compiler/rustc_mir_transform/src/prettify.rs:
```diff
@@ -0,0 +1,134 @@
+
+impl<'tcx> Visitor<'tcx> for LocalFinder {
+    fn visit_local(&mut self, l: Local, context: PlaceContext, _location: Location) {
+        if context.is_use() {
```
Why exclude non-uses?

scottmcm replied:
Because otherwise `StorageLive` ends up putting locals much earlier in the list than makes sense to me.  I'll comment it (and do the other suggestions).

@rustbot author

scottmcm committed:
PR feedback: better comments and debug asserts

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5038985705/jobs/9036832171) [(plain)](https://github.com/rust-lang/rust/commit/99ccad88aebc677da790053420ea71d90c514724/checks/13641804997/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_hir_typeck v0.1.0 (/checkout/compiler/rustc_hir_typeck)
    Checking rustc_privacy v0.0.0 (/checkout/compiler/rustc_privacy)
 Documenting rustc_privacy v0.0.0 (/checkout/compiler/rustc_privacy)
 Documenting rustc_hir_typeck v0.1.0 (/checkout/compiler/rustc_hir_typeck)
error: unresolved link to `BasicBlocks::successors`
  --> compiler/rustc_mir_transform/src/prettify.rs:16:37
   |
16 | /// Thus after this pass, all the [`BasicBlocks::successors`] of a block are
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^ the struct `BasicBlocks` has no field or associated item named `successors`
   |
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
error: could not document `rustc_mir_transform`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name rustc_mir_transform compiler/rustc_mir_transform/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/doc -Zunstable-options --check-cfg 'values(feature)' --check-cfg 'names()' --check-cfg 'values()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=b5ecc466144978c2 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps --extern either=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libeither-d57b38b997967b78.rmeta --extern itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libitertools-ad1fa586eb45f42f.rmeta --extern rustc_ast=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_ast-c26dfa863d5d54a5.rmeta --extern rustc_attr=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_attr-ac29d86dfe96a097.rmeta --extern rustc_const_eval=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_const_eval-d35854713bdcd385.rmeta --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-c29478f64130f4d1.rmeta --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-5a884c0fbab204f7.rmeta --extern rustc_fluent_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps/librustc_fluent_macro-d13f249698924795.so --extern rustc_hir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_hir-7f4691e632e9f80a.rmeta --extern rustc_index=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_index-fa4f8f1e19a83493.rmeta --extern rustc_macros=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps/librustc_macros-46578a62e90ab7e8.so --extern rustc_middle=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_middle-7ed5aad67adfca0c.rmeta --extern rustc_mir_dataflow=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_mir_dataflow-7860bbf2c4bbcb56.rmeta --extern rustc_serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_serialize-53d6c6a2b68c2078.rmeta --extern rustc_session=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_session-73b0f35a5e208c18.rmeta --extern rustc_span=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_span-a3a62e175a7e014e.rmeta --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-c3f55012650ae6f1.rmeta --extern rustc_trait_selection=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_trait_selection-cd5b5f2ab90b4be0.rmeta --extern smallvec=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsmallvec-4d2ed8dbf5f2b6bc.rmeta --extern tracing=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libtracing-98bd5c8b4fde9d1e.rmeta --extern-html-root-url 'either=https://docs.rs/either/1.6.0/' --extern-html-root-url 'itertools=https://docs.rs/itertools/0.10.5/' --extern-html-root-url 'smallvec=https://docs.rs/smallvec/1.10.0/' --extern-html-root-url 'tracing=https://docs.rs/tracing/0.1.35/' -Zunstable-options --cfg=bootstrap --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(parallel_compiler)' '--check-cfg=values(no_btreemap_remove_entry)' '--check-cfg=values(crossbeam_loom)' '--check-cfg=values(span_locations)' '--check-cfg=values(rustix_use_libc)' '--check-cfg=values(emulate_second_only_system)' '--check-cfg=values(windows_raw_dylib)' --document-private-items --document-hidden-items -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.71.0-nightly
  (c93d93859
  2023-05-21)' --document-private-items '-Arustdoc::private-intra-doc-links' --enable-index-page -Zunstable-options -Znormalize-docs --show-type-layout --generate-link-to-definition '-Zcrate-attr=warn(rust_2018_idioms)' --extern-html-root-url 'ena=https://docs.rs/ena/latest/'` (exit status: 1)
Build completed unsuccessfully in 0:00:39
```

</details>

scottmcm committed:
Normalize block and local orders in mir-opt tests

Since this only affects `PreCodegen MIR, and it would be nice for that to be resilient to permutations of things that don't affect the actual semantic behaviours.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5040398834/jobs/9039138651) [(plain)](https://github.com/rust-lang/rust/commit/131332511a3de468ff73878f4b570fe9bca978b8/checks/13644552458/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] tests/ui/borrowck/borrowck-thread-local-static-borrow-outlives-fn.rs stdout ----
diff of stderr:

4 LL |      assert_static(&FOO);
5    |                    ^^^^ thread-local variables cannot be borrowed beyond the end of the function
6 LL | }
-    | - end of enclosing function is here
+    |  - end of enclosing function is here
9 error: aborting due to previous error
10 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/borrowck/borrowck-thread-local-static-borrow-outlives-fn/borrowck-thread-local-static-borrow-outlives-fn.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args borrowck/borrowck-thread-local-static-borrow-outlives-fn.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/borrowck/borrowck-thread-local-static-borrow-outlives-fn.rs" "-Zthreads=1" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/borrowck/borrowck-thread-local-static-borrow-outlives-fn" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/borrowck/borrowck-thread-local-static-borrow-outlives-fn/auxiliary"
stdout: none
error[E0712]: thread-local variable borrowed past end of function
  --> fake-test-src-base/borrowck/borrowck-thread-local-static-borrow-outlives-fn.rs:8:20
   |
   |
LL |      assert_static(&FOO); //~ ERROR [E0712]
   |                    ^^^^ thread-local variables cannot be borrowed beyond the end of the function
LL | }
   |  - end of enclosing function is here
error: aborting due to previous error

For more information about this error, try `rustc --explain E0712`.
------------------------------------------
------------------------------------------


---- [ui] tests/ui/issues/issue-17954.rs stdout ----
diff of stderr:

5    |             ^^^^ thread-local variables cannot be borrowed beyond the end of the function
6 ...
7 LL | }
-    | - end of enclosing function is here
+    |  - end of enclosing function is here
10 error: aborting due to previous error
11 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-17954/issue-17954.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args issues/issue-17954.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/issues/issue-17954.rs" "-Zthreads=1" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-17954" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-17954/auxiliary"
stdout: none
error[E0712]: thread-local variable borrowed past end of function
  --> fake-test-src-base/issues/issue-17954.rs:7:13
   |
LL |     let a = &FOO;
LL |     let a = &FOO;
   |             ^^^^ thread-local variables cannot be borrowed beyond the end of the function
...
LL | }
   |  - end of enclosing function is here
error: aborting due to previous error

For more information about this error, try `rustc --explain E0712`.
------------------------------------------
---
diff of stderr:

2   --> $DIR/issue-52049.rs:6:10
3    |
4 LL |     foo(&unpromotable(5u32));
+    |     -----^^^^^^^^^^^^^^^^^^-- temporary value is freed at the end of this statement
6    |     |    |
7    |     |    creates a temporary value which is freed while still in use
7    |     |    creates a temporary value which is freed while still in use
8    |     argument requires that borrow lasts for `'static`
- LL | }
-    | - temporary value is freed at the end of this statement
11 
12 error: aborting due to previous error
---
To only update this specific test, also pass `--test-args issues/issue-52049.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/issues/issue-52049.rs" "-Zthreads=1" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-52049" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-52049/auxiliary"
stdout: none
error[E0716]: temporary value dropped while borrowed
  --> fake-test-src-base/issues/issue-52049.rs:6:10
   |
   |
LL |     foo(&unpromotable(5u32));
   |     -----^^^^^^^^^^^^^^^^^^-- temporary value is freed at the end of this statement
   |     |    creates a temporary value which is freed while still in use
   |     |    creates a temporary value which is freed while still in use
   |     argument requires that borrow lasts for `'static`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0716`.
------------------------------------------
------------------------------------------


---- [ui] tests/ui/lifetimes/issue-90600-expected-return-static-indirect.rs stdout ----
diff of stderr:

10    |                -------- cast requires that `foo` is borrowed for `'static`
12 LL | }
12 LL | }
-    | - `foo` dropped here while still borrowed
+    |  - `foo` dropped here while still borrowed
15 error: lifetime may not live long enough
16   --> $DIR/issue-90600-expected-return-static-indirect.rs:9:16



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lifetimes/issue-90600-expected-return-static-indirect/issue-90600-expected-return-static-indirect.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lifetimes/issue-90600-expected-return-static-indirect.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lifetimes/issue-90600-expected-return-static-indirect.rs" "-Zthreads=1" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lifetimes/issue-90600-expected-return-static-indirect" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lifetimes/issue-90600-expected-return-static-indirect/auxiliary"
stdout: none
error[E0597]: `foo` does not live long enough
  --> fake-test-src-base/lifetimes/issue-90600-expected-return-static-indirect.rs:7:32
   |
   |
LL | fn inner(mut foo: &[u8]) {
   |          ------- binding `foo` declared here
LL |     let refcell = RefCell::new(&mut foo);
   |                                ^^^^^^^^ borrowed value does not live long enough
LL |     //~^ ERROR `foo` does not live long enough
LL |     let read = &refcell as &RefCell<dyn Read>;
   |                -------- cast requires that `foo` is borrowed for `'static`
LL | }
LL | }
   |  - `foo` dropped here while still borrowed
error: lifetime may not live long enough
  --> fake-test-src-base/lifetimes/issue-90600-expected-return-static-indirect.rs:9:16
   |
   |
LL | fn inner(mut foo: &[u8]) {
   |                   - let's call the lifetime of this reference `'1`
...
LL |     let read = &refcell as &RefCell<dyn Read>;
   |                ^^^^^^^^ cast requires that `'1` must outlive `'static`
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0597`.
------------------------------------------
```

</details>

scottmcm committed:
MIR: opt-in normalization of `BasicBlock` and `Local` numbering

scottmcm commented on tests/mir-opt/inline/unchecked_shifts.rs:
```diff
@@ -4,6 +4,7 @@
 
 // ignore-debug: the debug assertions prevent the inlining we are testing for
 // compile-flags: -Zmir-opt-level=2 -Zinline-mir
+// compile-flags: -Zmir-enable-passes=+ReorderBasicBlocks,+ReorderLocals
```
Ooh, I like it, since it'll only affect tests that are using the `PreCodegen` MIR, not all the intermediate stuff.

scottmcm commented:
@rustbot ready

cjgillot commented on tests/ui/issues/issue-52049.stderr:
```diff
@@ -2,12 +2,10 @@ error[E0716]: temporary value dropped while borrowed
    |
 LL |     foo(&unpromotable(5u32));
-   |     -----^^^^^^^^^^^^^^^^^^-
+   |     -----^^^^^^^^^^^^^^^^^^-- temporary value is freed at the end of this statement
```
Why?

scottmcm replied:
Because I changed the RPO of
```
  A
 / \
B C
\ /
 D
```
from ` A C B D ` to `A B C D`.

I can pull that change out, if you'd prefer.

cjgillot commented:
Thanks! Those error are placed a bit better now anyway.
@bors r+

bors commented:
:pushpin: Commit d69725d5d5173e6c6fc20873f306cbd66fa963e7 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot d69725d5d5173e6c6fc20873f306cbd66fa963e7 -->
<!-- homu: {"type":"Approved","sha":"d69725d5d5173e6c6fc20873f306cbd66fa963e7","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d69725d5d5173e6c6fc20873f306cbd66fa963e7 with merge 089677eb32af83318467325edbef9b64053df532...
<!-- homu: {"type":"BuildStarted","head_sha":"d69725d5d5173e6c6fc20873f306cbd66fa963e7","merge_sha":"089677eb32af83318467325edbef9b64053df532"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5106572292/jobs/9179855938)
Approved by: cjgillot
Pushing 089677eb32af83318467325edbef9b64053df532 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5106572292/jobs/9179855938"},"merge_sha":"089677eb32af83318467325edbef9b64053df532"} -->

rust-timer commented:
Finished benchmarking commit (089677eb32af83318467325edbef9b64053df532): [comparison URL](https://perf.rust-lang.org/compare.html?start=1c53407e8c7cc922d718bde61ca34f47b6d2120f&end=089677eb32af83318467325edbef9b64053df532&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1c53407e8c7cc922d718bde61ca34f47b6d2120f&end=089677eb32af83318467325edbef9b64053df532&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.3%, 0.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.7%  | [0.3%, 1.2%]   | 8     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.0% | [-1.5%, -0.3%] | 7     |
| All ❌✅ (primary)                 | 0.3%  | [0.3%, 0.3%]   | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1c53407e8c7cc922d718bde61ca34f47b6d2120f&end=089677eb32af83318467325edbef9b64053df532&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1% | [0.1%, 0.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.3% | [2.3%, 2.3%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.1% | [0.1%, 0.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1c53407e8c7cc922d718bde61ca34f47b6d2120f&end=089677eb32af83318467325edbef9b64053df532&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.9% | [-1.9%, -1.9%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=1c53407e8c7cc922d718bde61ca34f47b6d2120f&end=089677eb32af83318467325edbef9b64053df532&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.0%, 0.6%]   | 50    |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 6     |
| Improvements ✅ <br /> (primary)   | -0.0% | [-0.0%, -0.0%] | 8     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-1.0%, -0.0%] | 9     |
| All ❌✅ (primary)                 | 0.2%  | [-0.0%, 0.6%]  | 58    |
</details>

**Bootstrap**: 646.44s -> 646.979s (0.08%)
